kotlin-inject-compiler/core/src/main/kotlin/me/tatarka/inject/compiler/TypeResultResolver.kt

```diff
@@ -61,30 +61,105 @@ class TypeResultResolver(private val provider: AstProvider, private val options:
+                        paramsWithName[param.name] = TypeResultRef(key, TypeResult.Arg(name))
+                    }
+                } else {
+                    assistedFailed = true
+                }
+            } else {
+                val result = resolveOrNull(context, element, key)
+                if (result != null) {
+                    paramsWithName[param.name] = result
+                } else if (!param.hasDefault) {
+                    throw FailedToGenerateException(cannotFind(key))
+                }
+            }
+        }
+
+        if (assistedFailed) {
+            throw FailedToGenerateException(
+                """
+                    Mismatched @Assisted parameters.
+                    Expected: [${
+                params.filter { it.isAssisted() }.joinToString()
+                }]"
+                    But got:  [${
+                args.joinToString { (type, _) -> type.toString() }
+                }"]
+                """.trimIndent(),
+                element
+            )
+        }
+        return paramsWithName
+    }
+
+    // old behavior where the last args are used, warn.
+    private fun resolveParamsLegacy(
+        context: Context,
+        element: AstElement,
+        params: List<AstParam>,
     ): Map<String, TypeResultRef> {
         val size = params.size
-        val args = context.args.asReversed()
         val paramsWithName = LinkedHashMap<String, TypeResultRef>(size)
-        params.forEachIndexed { i, param ->
+        val args = context.args.asReversed()
+        val resolvedImplicitly = mutableListOf<Pair<String, AstType>>()
+        for ((i, param) in params.withIndex()) {
             val indexFromEnd = size - i - 1
             val key = TypeKey(param.type, param.qualifier(options))
             val arg = args.getOrNull(indexFromEnd)
             if (arg != null) {
                 val (type, name) = arg
                 if (type.isAssignableFrom(key.type)) {
+                    resolvedImplicitly.add(param.name to key.type)
                     paramsWithName[param.name] = TypeResultRef(key, TypeResult.Arg(name))
-                    return@forEachIndexed
+                    continue
                 }
             }
             val result = resolveOrNull(context, element, key)
             if (result != null) {
                 paramsWithName[param.name] = result
-                return@forEachIndexed
-            }
-            if (!param.hasDefault) {
+            } else if (!param.hasDefault) {
                 throw FailedToGenerateException(cannotFind(key))
             }
         }
+
+        if (resolvedImplicitly.isNotEmpty()) {
+            provider.warn(
+                """
+                Implicit assisted parameters are deprecated and will be removed in a future version.
+                Annotate the following with @Assisted: [${
+                resolvedImplicitly.joinToString { (name, type) -> "$name: ${type.simpleName}" }
```

Comment:
Is it possible to reference line numbers? I had the experience a few times where a clickable line number would have been helpful to jump directly to the problematic code.

Reply:
By passing the element to the logger we should be getting clickable line numbers in the error message. If this is not working in some cases it would be worth looking into.